text/0000-cortex-m-targets.md

```diff
@@ -0,0 +1,608 @@
+- `data-layout`. (Code generation related). Also matches the value of the other ARM targets.
+- `executables`. Set to `true` because the Rust compiler can generate both libraries and executables
+  for these targets.
+- `os`. (For conditional compilation). No Rust precedent for this value as all the targets currently
+  supported by Rust have some form of OS or runtime. `none` has been chosen as this follows the
+  existing tooling (gcc and LLVM) convention of using `none` for the OS field of triples like
+  `arm-none-eabi`.
+- `pre-link-args` and `post-link-args`. By default, linkers are sensitive to the order of the
+  arguments they receive. Using these linker arguments undoes that default and can help avoid
+  "undefined reference" problems when linking statically. This technique is already used in the MUSL
+  targets that do static linking (`i686` and `x86_64` variants).
+- `target-endian`. (For conditional compilation). The Cortex-M architecture can be either little or
+  big endian. However, for microcontrollers, endianness is chosen at manufacture time and, as far as
+  the author knows, most (or all?) devices are little endian.
+- `target-pointer-width`. (For conditional compilation). All the existing Cortex-M processors are
+  32-bit processors.
+- `cpu`. (Enables CPU specific optimizations). The value pretty much matches the target name.
+- `features`. (Tweaks further the codege). On the Cortex-M4 and Cortex-M7 targets this tweaks the
+  use of FPU instructions. ARMv6-M based microcontrollers use `+strict-align` to prevent the
+  mis-optimization of the `copy_nonoverlapping` function.
+- `llvm-target`. (Controls code generation). Meaning of each field of the triple:
+  - `thumbv*m`. `thumb` indicates that the targets will use the Thumb instruction set instead of the
+    (older) ARM instruction set ("These processors support the Thumb instruction set only. -
+    [ARM]"). The `v*m` bit indicates the version/variant of the ARM architecture: `ARMv6-M`,
+    `ARMv7-M` or `ARMv7E-M`.
+  - `none`. Indicates the lack of an operating system.
+  - `eabi`. Indicates that the target uses the soft-float calling convention instead of the
+    hard-float calling convention (more details about this below).
+- `max-atomic-width`. ARMv6-M microcontrollers set this field to 0 to indicate that they don't
+  support atomics (there is no `dmb` instruction in their instruction set). This effectively removes
+  `AtomicI8` et al. from `core`.
+
+[ARM]: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.set.cortexm/index.html
+
+### Calling convention
+
+Targets that have a FPU can choose between two calling conventions (CC): *soft-float* or
+*hard-float*. Objects compiled with different CCs have different ABIs and can't always be linked
+together. Although the *hard-float* CC generates faster code around function calls that need to pass
+floating point arguments, this RFC proposes using the soft-float CC for these targets. The reason is
```

Comment:
It's mainly to avoid a explosion of the number of targets: if you have cortex-m4sf then you also need to add cortex-m4hf etc. The alternative proposal, which I prefer, does provides the two variants without adding some many targets to the compiler.